src/internal.rs

```diff
@@ -25,15 +44,67 @@ use core::sync::atomic::Ordering;
 use alloc::boxed::Box;
 
 use crossbeam_utils::cache_padded::CachePadded;
+use arrayvec::ArrayVec;
 
 use atomic::Owned;
 use collector::{Handle, Collector};
 use epoch::{AtomicEpoch, Epoch};
 use guard::{unprotected, Guard};
-use garbage::{Bag, Garbage};
+use deferred::Deferred;
 use sync::list::{List, Entry, IterError, IsElement};
 use sync::queue::Queue;
 
+/// Maximum number of objects a bag can contain.
+#[cfg(not(feature = "strict_gc"))]
+const MAX_OBJECTS: usize = 64;
+#[cfg(feature = "strict_gc")]
+const MAX_OBJECTS: usize = 4;
+
+/// Bag of garbages.
+#[derive(Default, Debug)]
+pub struct Bag {
+    /// Stashed objects.
+    deferreds: ArrayVec<[Deferred; MAX_OBJECTS]>,
+}
+
+// It is safe to implement `Send` for `Bag`, since the user of `Bag::try_push()` guarantees that
+// another thread may execute any `Deferred` functions inside a `Bag`.
+unsafe impl Send for Bag {}
+
+impl Bag {
+    /// Returns a new, empty bag.
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Returns `true` if the bag is empty.
+    pub fn is_empty(&self) -> bool {
+        self.deferreds.is_empty()
+    }
+
+    /// Attempts to insert a garbage object into the bag.
+    ///
+    /// Returns `Ok(())` if successful, and `Err(deferred)` for the given `deferred` if
+    /// unsuccessful.
+    ///
+    /// # Safety
+    ///
+    /// `deferred` should not have already been called. Also, another thread may execute the
```

Comment:
I don't think this actually affects safety. If you call a deferred function twice, the second call will simply panic.